network/src/connectivity_manager/test.rs

```diff
@@ -99,1204 +38,754 @@ fn gen_peer() -> (
     (peer_id, pubkey, pubkeys, addr)
 }
 
-async fn get_connected_size(conn_mgr_reqs_tx: &mut channel::Sender<ConnectivityRequest>) -> usize {
-    let (queue_size_tx, queue_size_rx) = oneshot::channel();
-    conn_mgr_reqs_tx
-        .send(ConnectivityRequest::GetConnectedSize(queue_size_tx))
-        .await
-        .unwrap();
-    queue_size_rx.await.unwrap()
+struct TestHarness {
+    trusted_peers: Arc<RwLock<PeerSet>>,
+    mock_time: MockTimeService,
+    connection_reqs_rx: diem_channel::Receiver<PeerId, ConnectionRequest>,
+    connection_notifs_tx: conn_notifs_channel::Sender,
+    conn_mgr_reqs_tx: channel::Sender<ConnectivityRequest>,
 }
 
-async fn get_dial_queue_size(conn_mgr_reqs_tx: &mut channel::Sender<ConnectivityRequest>) -> usize {
-    let (queue_size_tx, queue_size_rx) = oneshot::channel();
-    conn_mgr_reqs_tx
-        .send(ConnectivityRequest::GetDialQueueSize(queue_size_tx))
-        .await
-        .unwrap();
-    queue_size_rx.await.unwrap()
-}
+impl TestHarness {
+    fn new(
+        eligible_peers: Vec<PeerId>,
+        seed_addrs: HashMap<PeerId, Vec<NetworkAddress>>,
+    ) -> (Self, ConnectivityManager<FixedInterval>) {
+        ::diem_logger::Logger::init_for_testing();
 
-async fn send_new_peer_await_delivery(
-    connection_notifs_tx: &mut conn_notifs_channel::Sender,
-    peer_id: PeerId,
-    notif_peer_id: PeerId,
-    address: NetworkAddress,
-) {
-    let mut metadata = ConnectionMetadata::mock(notif_peer_id);
-    metadata.addr = address;
-    let notif = peer_manager::ConnectionNotification::NewPeer(metadata, NetworkContext::mock());
-    send_notification_await_delivery(connection_notifs_tx, peer_id, notif).await;
-}
-
-async fn send_lost_peer_await_delivery(
-    connection_notifs_tx: &mut conn_notifs_channel::Sender,
-    peer_id: PeerId,
-    notif_peer_id: PeerId,
-    address: NetworkAddress,
-    reason: DisconnectReason,
-) {
-    let mut metadata = ConnectionMetadata::mock(notif_peer_id);
-    metadata.addr = address;
-    let notif =
-        peer_manager::ConnectionNotification::LostPeer(metadata, NetworkContext::mock(), reason);
-    send_notification_await_delivery(connection_notifs_tx, peer_id, notif).await;
-}
+        // Fills in the gaps for eligible peers. Eligible peers in tests aren't
+        // always the seeds.
+        let seeds: PeerSet = eligible_peers
+            .into_iter()
+            .map(|peer_id| {
+                let addrs: Vec<NetworkAddress> = seed_addrs
+                    .get(&peer_id)
+                    .map_or_else(Vec::new, |addr| addr.clone());
+                let mut pubkeys = addrs
+                    .iter()
+                    .filter_map(NetworkAddress::find_noise_proto)
+                    .collect::<HashSet<_>>();
+                if pubkeys.is_empty() {
+                    pubkeys.insert(x25519::PrivateKey::generate_for_testing().public_key());
+                }
+                (peer_id, Peer::new(addrs, pubkeys, PeerRole::Validator))
```

Comment:
Do you expect these Addrs to have pubkeys in them?  I expected that test developers would be too lazy, and it would generate the key instead.

Reply:
gen_peer creates keys i believe

the tests could be more consistent about this... but I was too tired to change it lol